onnxruntime/core/graph/graph.cc

```diff
@@ -380,13 +412,64 @@ common::Status NodeArg::UpdateTypeAndShape(const ONNX_NAMESPACE::TypeProto& inpu
           *current_type.mutable_sparse_tensor_type() = input_tensor_type;
         }
       }
-
       break;
     }
 #endif
+
+    case TypeProto::kOptionalType: {
+      if (utils::HasOptionalTensorType(input_type) && !utils::HasOptionalTensorType(current_type)) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Optional Type mismatch. Expected: Seq tensor. Got: Tensor");
+      } else if (!utils::HasOptionalTensorType(input_type) && utils::HasOptionalTensorType(current_type)) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Optional Type mismatch. Expected: Tensor. Got: Seq Tensor");
+      }
+
+      // Updating element type and shape is only applicable for optional tensors
+      if (utils::HasOptionalTensorType(input_type)) {
+        const auto& optional_input_type = utils::GetOptionalTypeProto(input_type);
+        auto& optional_current_type = *utils::GetMutableOptionalTypeProto(current_type);
+
+        const auto& input_tensor_type = optional_input_type.tensor_type();
+        const auto& input_tensor_elem_type = input_tensor_type.elem_type();
+        const auto& current_tensor_elem_type = optional_current_type.tensor_type().elem_type();
+
+        if (input_tensor_elem_type != current_tensor_elem_type) {
+          if (override_types) {
+            DataType inferred_type = DataTypeUtils::ToType(input_type);
+            // The "SetType" call will override the shape information to empty.
+            // If the original tensor has shape information, need to set it back.
+            if (Shape()) {
+              auto old_shape = *Shape();
+              SetType(inferred_type);
+              SetShape(old_shape);
+            } else {
+              SetType(inferred_type);
+            }
+          } else {
+            return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Optional Tensor element type mismatch. ",
+                                   static_cast<TensorProto_DataType>(input_tensor_elem_type), " != ",
+                                   static_cast<TensorProto_DataType>(current_tensor_elem_type));
+          }
+        }
+
+        if (utils::HasShape(optional_input_type.tensor_type())) {
+          if (utils::HasShape(optional_current_type.tensor_type())) {
+            ORT_RETURN_IF_ERROR(MergeShapeInfo(Name(), optional_input_type, optional_current_type, strict, logger));
+          } else {
+            *optional_current_type.mutable_tensor_type() = input_tensor_type;
+          }
+        }
+      } else {
+        // TODO: What is the plan for optional sequence tensors ?
+        // Currently, we don't do anything for the generic sequence type
+        // as seen below. This section needs an update if we choose to
+        // change that in the future.
+      }
```

Comment:
Would an INFO level message would be useful or do we know it's completely valid to do nothing here?

Reply:
Today there is nothing being done for the sequence type as the comment reflects. Not sure what INFO level logging would be useful to the user

Reply:
Was just wondering if we could hit a scenario when doing nothing caused the model to fail to load. If that happened there would be no hints as to why.